### PR TITLE
Structs as args

### PR DESCRIPTION
**Pull request by charles-cooper**
### - What I did
Partially implement #1019 - structs as arguments and return values.

### WIP Checklist
- [x] Fix ABI output for function args
- [x] Fix units
- [x] Fix ABI output for nested structs/tuples
- [x] Tests

### - How I did it
Add packing / unpacking code for structs

### - How to verify it
See tests

### - Description for the changelog
Ability to pass structs as args and return values. Structs which contain other structs or structs which contain bytearrays not supported.

### CODE REVIEW

File: vyper/parser/expr.py [hunk 1]
```diff
@@ -239,7 +239,7 @@ def attribute(self):
                 return sub
             if not isinstance(sub.typ, StructType):
                 raise TypeMismatchException("Type mismatch: member variable access not expected", self.expr.value)
-            attrs = sorted(sub.typ.members.keys())
+            attrs = list(sub.typ.members.keys())
```
Comment (jacqueswww) [inline vyper/parser/expr.py]: Why remove sorted?

Comment (jacqueswww) [reply vyper/parser/expr.py]: ah nvm, I see.

File: vyper/parser/expr.py [hunk 2]
```diff
@@ -674,26 +674,53 @@ def call(self):
...
+                #    raise StructureException("Struct constructor must be called in RHS of assignment.", self.expr)
+                args = self.expr.args
+                if len(args) != 1:
+                    raise StructureException("Struct constructor is called with one argument only")
+                arg = args[0]
+                # Only allow construction if argument is a dict value
+                sub = Expr.parse_value_expr(arg, self.context)
+                if not (isinstance(sub.typ, StructType)):
+                    if sub.type is not None:
+                        raise StructureException("Struct be constructed with a dict")
```
Comment (jacqueswww) [inline vyper/parser/expr.py]: add self.expr to this exception.

File: vyper/parser/expr.py [hunk 3]
```diff
@@ -674,26 +674,53 @@ def call(self):
...
                 return dispatch_table[function_name](self.expr, self.context)
+
+            # Struct constructors do not need `self` prefix.
+            elif function_name in self.context.structs:
+                # Can also be called in return statement...
+                #if not self.context.in_assignment and not self.context.:
+                #    raise StructureException("Struct constructor must be called in RHS of assignment.", self.expr)
+                args = self.expr.args
+                if len(args) != 1:
+                    raise StructureException("Struct constructor is called with one argument only")
```
Comment (jacqueswww) [inline vyper/parser/expr.py]: add self.expr to this exception

File: vyper/parser/expr.py [hunk 4]
```diff
@@ -674,26 +674,53 @@ def call(self):
...
             function_name = self.expr.func.id
+
             if function_name in dispatch_table:
                 return dispatch_table[function_name](self.expr, self.context)
+
+            # Struct constructors do not need `self` prefix.
+            elif function_name in self.context.structs:
+                # Can also be called in return statement...
+                #if not self.context.in_assignment and not self.context.:
+                #    raise StructureException("Struct constructor must be called in RHS of assignment.", self.expr)
```
Comment (jacqueswww) [inline vyper/parser/expr.py]: why was this exception commented out?

Comment (charles-cooper) [reply vyper/parser/expr.py]: Struct constructors can also be called in return statements

**Comment by charles-cooper** [comment]
Not in this PR. I think that structs in events needs a bit more thought. Since each byte costs 8 gas, we probably should use the packed ABI encoding. I'm not sure what solidity produces / what web3 expects.

**Comment by fubuloubu** [comment]
Moved that convo to #1170

**Comment by charles-cooper** [comment]
Ok, based on the latest meeting, I rebased this PR and reduced the scope. For now, nested structs will not work with the ABI, and neither will struct arguments. Also, I opted to disallow passing structs with dynamic members for now, otherwise this PR will get too large. Should be ready for review (pending CI)!

**Comment by fubuloubu** [comment]
CI failed, but I also think this feature needs way more tests lol

**Comment by fubuloubu** [comment]
Also, can you add issues to track what was deemed out of scope?

**Comment by charles-cooper** [comment]
@fubuloubu I can add some typechecking tests. Any other specific tests you think should be in here?

**Comment by fubuloubu** [comment]
Anything with the ABI should be pretty throughly tested. That boundary a lot of things can go wrong haha.

**Comment by fubuloubu** [comment]
@jacqueswww when merged, can you cut as 0.1.0b7?

**Comment by jacqueswww** [comment]
@fubuloubu I want a few more bug fixes in first, but b7 will be soon. In the meantime just peg the commit hash in your requirements.

**Comment by fubuloubu** [comment]
Testing this. It compiles, but I'm having difficulties getting the ABI to work with Web3py. What's the best way to do this?

**Comment by jacqueswww** [comment]
@fubuloubu I don't think web3py is ready for those types of nested tuples (yet).

**Comment by jacqueswww** [comment]
I think I will just merge this for now - as I am scared of branches diverging too much. Will create a ticket for the outstanding test.

**Comment by fubuloubu** [comment]
Follow-up: ABI v2 doesn't seem to work with this :(

**Comment by charles-cooper** [comment]
@fubuloubu uh oh! Can you provide a reproducible case?

**Comment by fubuloubu** [comment]
dependencies:
`pip install "py-evm>=0.2.0b39"`
`pip install "eth-tester>=0.1.0b36"`
`pip install "web3>=5.0.0a2"`
`pip install git+https://github.com/ethereum/vyper.git`

test.py:
```python
>>> import vyper
>>> from web3 import Web3, EthereumTesterProvider

>>> w3 = Web3(EthereumTesterProvider())

>>> interface = vyper.compile_code("""
... struct MyStruct:
...     a: address
...     b: bytes32
... 
... @public
... def a(s: MyStruct) -> address:
...     return s.a
... 
... @public
... def b(s: MyStruct) -> bytes32:
...     return s.b
... """, output_formats=['abi', 'bytecode', 'bytecode_runtime'])

>>> txn_hash = w3.eth.contract(**interface).constructor().transact()
>>> address = w3.eth.waitForTransactionReceipt(txn_hash)['contractAddress']
>>> contract = w3.eth.contract(address, **interface)

>>> contract.functions.a(contract.address, b'\x00' * 32).call()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/bryant/.pyenv/versions/test/lib/python3.6/site-packages/web3/contract.py", line 1003, in __call__
    clone._set_function_info()
  File "/home/bryant/.pyenv/versions/test/lib/python3.6/site-packages/web3/contract.py", line 1012, in _set_function_info
    self.kwargs
  File "/home/bryant/.pyenv/versions/test/lib/python3.6/site-packages/web3/_utils/contracts.py", line 128, in find_matching_fn_abi
    raise ValidationError(message)
web3.exceptions.ValidationError:
Could not identify the intended function with name `a`, positional argument(s) of type `(<class 'str'>, <class 'bytes'>)` and keyword argument(s) of type `{}`.
Found 1 function(s) with the name `a`: ['a((address,bytes32))']
Function invocation failed due to improper number of arguments.

>>> contract.functions.a((contract.address, b'\x00' * 32)).call()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/bryant/.pyenv/versions/test/lib/python3.6/site-packages/web3/contract.py", line 1003, in __call__
    clone._set_function_info()
  File "/home/bryant/.pyenv/versions/test/lib/python3.6/site-packages/web3/contract.py", line 1012, in _set_function_info
    self.kwargs
  File "/home/bryant/.pyenv/versions/test/lib/python3.6/site-packages/web3/_utils/contracts.py", line 97, in find_matching_fn_abi
    function_candidates = pipe(abi, *filters)
  File "cytoolz/functoolz.pyx", line 589, in cytoolz.functoolz.pipe
  File "cytoolz/functoolz.pyx", line 565, in cytoolz.functoolz.c_pipe
  File "/home/bryant/.pyenv/versions/test/lib/python3.6/site-packages/web3/_utils/abi.py", line 203, in filter_by_encodability
    in contract_abi
  File "/home/bryant/.pyenv/versions/test/lib/python3.6/site-packages/web3/_utils/abi.py", line 204, in <listcomp>
    if check_if_arguments_can_be_encoded(function_abi, args, kwargs)
  File "/home/bryant/.pyenv/versions/test/lib/python3.6/site-packages/web3/_utils/abi.py", line 221, in check_if_arguments_can_be_encoded
    for _type, arg in zip(types, arguments)
  File "/home/bryant/.pyenv/versions/test/lib/python3.6/site-packages/web3/_utils/abi.py", line 221, in <genexpr>
    for _type, arg in zip(types, arguments)
  File "/home/bryant/.pyenv/versions/test/lib/python3.6/site-packages/web3/_utils/abi.py", line 166, in is_encodable
    base, sub, arrlist = process_type(_type)
  File "/home/bryant/.pyenv/versions/test/lib/python3.6/site-packages/web3/_utils/abi.py", line 136, in process_type
    type_str_repr,
ValueError: Cannot process type '(address,bytes32)': tuple types not supported
```

**Comment by fubuloubu** [comment]
A think it's a bug against web3py v5 if it doesn't implement the full abi V2 spec yet.

How did you decide what the ABI output should be in moving forward with this PR? Basically follow Solidity?

**Comment by charles-cooper** [comment]
@fubuloubu Oh that's right. I even said so in my comments above (and apparently promptly forgot!)

To answer your question, I wanted to follow Solidity but tuples are not supported by eth-abi yet. The signature should be ABI-encoded as
```
...
    "name": "b",
    "inputs": [
      {
        "type": "tuple",
        "components": [
          { "type": "address", "name": "s_a"},
          {"type": "bytes32", "name" "s_b"}],
        "name": "s"
      }
    ],
...
```
and something similar for outputs. But that doesn't work, so for now I kept the input/output encoding the same as before (tuple inputs are encoded as `(address,bytes32)` and outputs are encoded as an array `"outputs": [{ "type": "address", "name": "s_a"},{"type": "bytes32", "name" "s_b"}]`) so as to keep tests passing.

**Comment by fubuloubu** [comment]
I really wished the ABI had entries for structs, so we can just reference structs in different places instead of this tuple nonsense lol

**Pull request merged**